### PR TITLE
http: applying route header rules on local replies.

### LINKED ISSUE #4317: Envoy doesn't add headers in case of Gateway Timeout
*Description*:

Envoy doesn't add headers specified in the route configuration in case of upstream timeout.

*Repro steps*:

Send request to upstream which doesn't respond within specified route timeout.

Expected: envoy adds headers specified under `response_headers_to_add`

Actual: envoy doesn't add any headers:

```
          'server' => 'envoy',
          'content-length' => '24',
          'Reason' => 'Gateway Timeout',
          'URL' => 'http://authxserver.service/infra_test/idempotent',
          'HTTPVersion' => '1.1',
          'content-type' => 'text/plain',
          'date' => 'Fri, 31 Aug 2018 13:21:31 GMT',
          'Status' => 504
```

*route config*:

```
    {
     "match": {
      "regex": "^/(infra_test/idempotent|/iam/issue-refresh-token|identity/user-exists/.*)"
     },
     "route": {
      "cluster": "authxserver.prod.cluster",
      "timeout": "0.500s",
      "retry_policy": {
       "retry_on": "5xx",
       "num_retries": 2,
       "per_try_timeout": "0.150s"
      },
      "response_headers_to_add": [
       {
        "header": {
         "key": "x-envoy-authxserver-endpoint-type",
         "value": "3"
        }
       }
      ]
     }
```

### PR DESCRIPTION
**Pull request by alyssawilk**
Commit Message: http: applying route header rules on local replies.
Risk Level: Medium (L7 change)
Testing: updated integration tests
Docs Changes: n/a
Release Notes: inline
Fixes #4317
Runtime Guard: envoy.reloadable_features.always_apply_route_header_rules

### CODE REVIEW

File: source/common/http/filter_manager.cc [hunk 1]
```diff
@@ -812,7 +812,16 @@ void FilterManager::sendLocalReplyViaFilterChain(
   Utility::sendLocalReply(
       state_.destroyed_,
       Utility::EncodeFunctions{
-          modify_headers,
+          [this, modify_headers](ResponseHeaderMap& headers) -> void {
+            if (streamInfo().route_entry_ &&
```
Comment (yanavlasov) [inline source/common/http/filter_manager.cc:816]: Is it possible to make a test for the case where `route_entry_` is nullptr?

Comment (alyssawilk) [reply source/common/http/filter_manager.cc:816]: every early response test (invalid HTTP/1.1 sending 400 before we have headers and can select route) tests this.  Did you want a unit test as well, or just to make sure we had coverage?

File: source/common/http/filter_manager.cc [hunk 2]
```diff
@@ -846,7 +855,16 @@ void FilterManager::sendDirectLocalReply(
...
+                Runtime::runtimeFeatureEnabled(
+                    "envoy.reloadable_features.always_apply_route_header_rules")) {
+              streamInfo().route_entry_->finalizeResponseHeaders(headers, streamInfo());
+            }
+            if (modify_headers) {
+              modify_headers(headers);
+            }
```
Comment (mattklein123) [inline source/common/http/filter_manager.cc:866]: This stanza is duplicated verbatim from above and is a little tricky so might be worth de-deduping but up to you.

**Review by mattklein123 — APPROVED** [review]
Awesome, thanks for fixing this long standing issue! nice to see the fix be so simple after all of the various refactors in this code path recently.

File: source/common/http/filter_manager.cc (hunk 1 above)
Comment (yanavlasov) [reply source/common/http/filter_manager.cc:816]: Just wanted to make sure we have coverage

**Comment by mike1808** [comment]
Hello. I'm not familiar with your backport policy and I'm curious if are you going to backport this bug to 1.5?

**Comment by alyssawilk** [comment]
We can cc @envoyproxy/stable-maintainers and ask them what they think!

**Comment by PiotrSikora** [comment]
Since this is neither security nor stability fix, it doesn't qualify for backporting.
See: https://github.com/envoyproxy/envoy/blob/master/RELEASES.md#stable-releases

**Comment by alyssawilk** [comment]
Right, the question more if we consider it an enhancement, or a bugfix fixing inconsistent header handling.

**Pull request merged**